tools/serverpod_cli/lib/src/migrations/generator.dart

```diff
@@ -264,96 +208,143 @@ class MigrationGenerator {
+    var installedModules = liveDatabase.installedModules.where(
+      (module) => module.module != MigrationConstants.repairMigrationModuleName,
+    );
 
-    installedModules.removeWhere((module) =>
-        module.module == MigrationConstants.repairMigrationModuleName);
+    var targetModules = targetDatabase.installedModules.where(
+      (module) => module.module != MigrationConstants.repairMigrationModuleName,
+    );
 
-    if (installedModules.length != versions.length) {
+    if (installedModules.length != targetModules.length) {
       return true;
     }
 
     for (var module in installedModules) {
-      if (versions[module.module]?.versionName != module.version) {
+      if (targetModules.any((version) =>
+          version.module == module.module &&
+          version.version != module.version)) {
         return true;
       }
     }
 
     return false;
   }
 
-  DatabaseDefinition createDatabaseDefinitionFromTables(
-    Map<String, MigrationVersion> versions,
-  ) {
-    var migrationDefinitions =
-        versions.values.map((e) => e.databaseDefinition).toList();
+  Future<List<MigrationVersion>> _loadMigrationVersionsFromModules(
+    List<ModuleConfig> modules, {
+    String? targetMigrationVersion,
+    Directory? directory,
+  }) async {
+    var modulePaths = await locateAllModulePaths(
+      directory: directory ?? Directory.current,
+    );
 
-    var installedModules = versions.values
-        .map((value) => DatabaseMigrationVersion(
-            module: value.moduleName, version: value.versionName))
-        .toList();
+    var selectedModules = modules.where(
+      (module) => module.migrationVersions.isNotEmpty,
+    );
 
-    var dstDatabase = DatabaseDefinition(
-      tables: migrationDefinitions.expand((e) => e.tables).toList(),
-      migrationApiVersion: DatabaseConstants.migrationApiVersion,
-      installedModules: installedModules,
+    var selectedPaths = modulePaths.where(
+      (modulePath) {
+        var moduleName = _extractModuleNameFromPath(modulePath);
+        return selectedModules.any((module) => module.name == moduleName);
+      },
     );
-    return dstDatabase;
-  }
 
-  Future<Map<String, MigrationVersion>> loadMigrationVersionsFromAllModules({
-    RepairTargetMigration? targetMigration,
-  }) async {
-    var versions = <String, MigrationVersion>{};
-    if (targetMigration != null) {
-      versions[targetMigration.moduleName] =
-          await _loadTargetRepairMigrationVersion(targetMigration);
+    var moduleMigrationVersions = <MigrationVersion>[];
+
+    for (var modulePath in selectedPaths) {
+      var moduleName = _extractModuleNameFromPath(modulePath);
+      var versionName = selectedModules
+          .firstWhere((e) => e.name == moduleName)
+          .migrationVersions
+          .last;
+
+      var migrationVersion = await MigrationVersion.load(
+        moduleName: moduleName,
+        versionName: targetMigrationVersion ?? versionName,
```

Comment:
Quesiton: Hmm, how does this work with `targetMigrationVersion` parameter? Will we load the same target migration version for all modules in that case?

Reply:
I don't understand this. Isn't the purpose of this method to load the migration version for all modules passed in? In that case the target migration would be used for all those modules if I understand the code correctly.